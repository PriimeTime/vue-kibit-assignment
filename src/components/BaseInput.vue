<script setup lang="ts">
import type { InputType } from "@/types/InputType";

const props = defineProps({
  modelValue: {
    type: String,
    required: false,
  },
  type: {
    type: String,
    default: "text" as InputType,
  },
  placeholder: {
    type: String,
    required: false,
  },
});

const emit = defineEmits(["update:modelValue", "keydown"]);
</script>

<template>
  <div>
    <template v-if="props.type === 'textarea'">
      <textarea
        class="border-teal-900 border-2 rounded-lg p-2 w-full"
        :placeholder="props.placeholder"
        :value="props.modelValue"
        @keydown="emit('keydown', $event)"
        @input="
          emit(
            'update:modelValue',
            ($event.target as HTMLTextAreaElement)?.value
          )
        "
      ></textarea>
    </template>
    <template v-else>
      <input
        class="border-teal-900 border-2 rounded-lg p-2 w-full"
        :type="props.type"
        :placeholder="props.placeholder"
        :value="props.modelValue"
        @keydown="emit('keydown', $event)"
        @input="
          emit('update:modelValue', ($event.target as HTMLInputElement)?.value)
        "
      />
    </template>
  </div>
</template>
