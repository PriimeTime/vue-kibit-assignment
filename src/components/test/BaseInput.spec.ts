import { describe, it, expect, vi } from "vitest";
import { mount } from "@vue/test-utils";
import BaseInput from "@/components/BaseInput.vue";

describe("BaseInput", () => {
  it("renders as an input by default", () => {
    const wrapper = mount(BaseInput, {
      props: {
        modelValue: "",
        placeholder: "Enter text",
      },
    });
    expect(wrapper.find("input").exists()).toBe(true);
  });

  it("renders as a textarea when type is 'textarea'", () => {
    const wrapper = mount(BaseInput, {
      props: {
        modelValue: "",
        type: "textarea",
        placeholder: "Enter text",
      },
    });
    expect(wrapper.find("textarea").exists()).toBe(true);
  });

  it("emits 'update:modelValue' when input changes", async () => {
    const wrapper = mount(BaseInput, {
      props: {
        modelValue: "",
        placeholder: "Enter text",
      },
    });
    const input = wrapper.find("input");
    await input.setValue("Hello World");
    const emitted = wrapper.emitted("update:modelValue");
    expect(emitted).toBeTruthy();
    expect(emitted?.[0]).toEqual(["Hello World"]);
  });

  it("emits 'keydown' when a key is pressed", async () => {
    const wrapper = mount(BaseInput, {
      props: {
        modelValue: "",
        placeholder: "Enter text",
      },
    });
    const input = wrapper.find("input");
    await input.trigger("keydown", { key: "Enter" });

    const emitted = wrapper.emitted("keydown");
    expect(emitted).toBeTruthy();
    expect(emitted?.[0][0]).toHaveProperty("key", "Enter");
  });

  it("matches snapshot for input", () => {
    const wrapper = mount(BaseInput, {
      props: {
        modelValue: "",
        placeholder: "Enter text",
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("matches snapshot for textarea", () => {
    const wrapper = mount(BaseInput, {
      props: {
        modelValue: "",
        type: "textarea",
        placeholder: "Enter text",
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
});
