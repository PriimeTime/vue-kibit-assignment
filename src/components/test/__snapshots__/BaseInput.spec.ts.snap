// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BaseInput > matches snapshot for input 1`] = `"<div><input class="border-teal-900 border-2 rounded-lg p-2 w-full focus:outline-none" placeholder="Enter text" value=""></div>"`;

exports[`BaseInput > matches snapshot for textarea 1`] = `"<div><textarea class="border-teal-900 border-2 rounded-lg p-2 w-full focus:outline-none" placeholder="Enter text" value=""></textarea></div>"`;
